### PR TITLE
Add conflict detection feature to apply strategy

### LINKED ISSUE #60945: Add conflict detection feature to apply strategy
**FEATURE REQUEST**:
>
> /kind feature
    Implement feature for https://github.com/kubernetes/kubectl/issues/97

**What is expected to happen**:
Apply is being rewritten under pkg/kubectl/apply/strategy based on visitor pattern. The new merge and replace code should check for conflicts between the recorded value and the remote value, and optionally return an error if they do not match with the field and details. A conflict is if the same field is specified in BOTH the recorded and the remote values of an object, but does not match.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.10
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by fanzhangio**
- Add DetectConflictor interface on Element level. Implemented it for particular elements.
- If Options.FailOnConflict is enabled, Merge will detect conflict by invoking doConflictDecect for particular element,
  returning ConflictError with details.
- Add tests, including use case examples and illustration. For example: list, map, and complicated combination.

**What this PR does / why we need it**:
Apply is being rewritten under pkg/kubectl/apply/strategy based on visitor pattern. The new merge and replace code should check for conflicts between the recorded value and the remote value, and optionally return an error if they do not match with the field and details. A conflict is if the same field is specified in BOTH the recorded and the remote values of an object, but does not match.

**Which issue(s) this PR fixes**:
Fixes #60945 
https://github.com/kubernetes/kubectl/issues/97

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/apply/strategy/replace_visitor.go [hunk 1]
```diff
@@ -98,3 +109,42 @@ func (v replaceStrategy) doAddOrDelete(e apply.Element) (apply.Result, bool) {
 }
 
 var _ apply.Strategy = &replaceStrategy{}
+
+func (v replaceStrategy) detectConflict(e apply.Element) (apply.Result, error) {
+	switch e.(type) {
+	case apply.ListElement:
```
Comment (pwittrock) [inline pkg/kubectl/apply/strategy/replace_visitor.go]: Why isn't this a function on the specific element type?

Comment (pwittrock) [inline pkg/kubectl/apply/strategy/replace_visitor.go]: The behavior for each of these should be documented.  What does it mean to detect a conflict for a list?  It shouldn't be just if the elements have conflicts, it should also check if there are any elements missing or added.

File: pkg/kubectl/apply/visitor.go [hunk 2]
```diff
@@ -46,6 +46,8 @@ const (
 	SET
 	// DROP drops the field from an object
 	DROP
+	// CONFLICT occurs in recorded and remote
+	CONFLICT
```
Comment (pwittrock) [inline pkg/kubectl/apply/visitor.go]: Why do we want this over just returning ERROR?

File: pkg/kubectl/apply/strategy/replace_visitor.go [hunk 3]
```diff
@@ -65,7 +74,9 @@ func (v replaceStrategy) MergeEmpty(e apply.EmptyElement) (apply.Result, error)
 // replace returns the local value if specified, otherwise it returns the remote value
 // this works regardless of the approach
 func (v replaceStrategy) doReplace(e apply.Element) (apply.Result, error) {
-	// TODO: Check for conflicts
+	if v.options.FailOnConflict {
```
Comment (pwittrock) [inline pkg/kubectl/apply/strategy/replace_visitor.go]: We need to have very good documentation for what constitutes a conflict in `FailOnConflict `.  How does it behave for lists, maps, primitives?

File: pkg/kubectl/apply/strategy/replace_visitor.go [hunk 4]
```diff
@@ -53,7 +55,14 @@ func (v replaceStrategy) MergeType(e apply.TypeElement) (apply.Result, error) {
 
 // MergePrimitive returns a result by merging the recorded, local and remote values
 // - replacing the remote value with the local value
+// If FailOnConflict option is enabled, throw error when conflicts occure in both recorded
+// and remote values.
 func (v replaceStrategy) MergePrimitive(e apply.PrimitiveElement) (apply.Result, error) {
+	if v.options.FailOnConflict && e.HasRecorded() && e.HasRemote() &&
```
Comment (pwittrock) [inline pkg/kubectl/apply/strategy/replace_visitor.go]: This is a really complex conditional.  How can we structure this to make it clear what the intent is?

File: pkg/kubectl/apply/strategy/replace_visitor.go [hunk 5]
```diff
@@ -53,7 +55,14 @@ func (v replaceStrategy) MergeType(e apply.TypeElement) (apply.Result, error) {
...
+		fmt.Sprintf("%v", e.GetRecorded()) != fmt.Sprintf("%v", e.GetRemote()) {
+		return apply.Result{Operation: apply.CONFLICT}, fmt.Errorf("Conflict detected in recorded value (%+v) and remote value (%+v)",
```
Comment (pwittrock) [inline pkg/kubectl/apply/strategy/replace_visitor.go]: Does this contain the full path to the conflicting value?  If not it will be hard for users to debug.  Don't fix now if it is going to add a bunch of code, but we should at least add a TODO for how to fix this.

File: pkg/kubectl/apply/strategy/merge_visitor.go [hunk 6]
```diff
@@ -99,6 +100,9 @@ func (v mergeStrategy) doMergeMap(e map[string]apply.Element) (apply.Result, err
 		// Recursively merge the map element before adding the value to the map
 		result, err := value.Merge(v.strategic)
 		if err != nil {
+			if result.Operation == apply.CONFLICT {
```
Comment (pwittrock) [inline pkg/kubectl/apply/strategy/merge_visitor.go]: Why do we need this?  Can we just return the result, or even not have the CONFLICT state.

File: pkg/kubectl/apply/visitor.go (hunk 2 above)
Comment (pwittrock) [reply pkg/kubectl/apply/visitor.go]: Instead of doing this, we should have different error subclasses with one for conflict so the user can check the error type.

File: pkg/kubectl/apply/strategy/replace_visitor.go (hunk 1 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/strategy/replace_visitor.go]: I originally wanted to add this small function in strategy without impact Element interface and impl. I got your point, so I want to do as following:
1. Add an interface `DetectConflict(Strategy)` in apply/Element.go
2. Implement this interface for each particular element.
Does it make sense ?

File: pkg/kubectl/apply/strategy/replace_visitor.go (hunk 5 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/strategy/replace_visitor.go]: Good point. I have taken it into consideration. 
primitiveElement is the 'leaf' node for comparing, if conflicts occured, the error message will be thrown out stack. Error message contains the conflict values. In call stack, I want to append the current node with the cached error message recursively, so that the path will be presented.  BTW, the `CONFLICT` is used for identification, without particular error class for conflict by far.

File: pkg/kubectl/apply/strategy/replace_visitor.go (hunk 1 above)
Comment (pwittrock) [reply pkg/kubectl/apply/strategy/replace_visitor.go]: The simplest thing may be to add the interface for conflict detection and then check if e implements the interface - then call the function.

File: pkg/kubectl/apply/visitor.go (hunk 2 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/visitor.go]: CONFLICT is used as identification by far, since we have not particular error class for conflict. If we are going to add one, I can remove it.

**Review by droot — CHANGES_REQUESTED** [review]
@fanzhangio This definitely looks better than the previous approach where we had compare-strategy :)

I agree with Phil's comment and we should address the following areas of concern:
 - Define ConflictError in apply package and return that during Merge instead of implementing conflict as another operation in Result.
 - Implement conflict detection at the element level to avoid the monolithic switch-case
 - Improve the documentation.

File: pkg/kubectl/apply/primitive_element.go [hunk 7]
```diff
@@ -32,3 +36,27 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
...
+	}
+	return Result{}, nil
+}
+
+func checkPrimitive(e PrimitiveElement) bool {
+	if !e.HasRecorded() && !e.HasRemote() {
+		return true
+	}
+	if e.HasRecorded() && e.HasRemote() {
+		return fmt.Sprintf("%v", e.GetRecorded()) == fmt.Sprintf("%v", e.GetRemote())
```
[the remark this answers is not in the document]
Comment (dougm) [reply pkg/kubectl/apply/primitive_element.go]: `reflect.DeepEqual` ?

File: pkg/kubectl/apply/primitive_element.go [hunk 8]
```diff
@@ -32,3 +36,27 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
 }
 
 var _ Element = &PrimitiveElement{}
+
+// ConflictDetect implement for primitive element.
```
Comment (dougm) [inline pkg/kubectl/apply/primitive_element.go]: implement -> implemention

File: pkg/kubectl/apply/primitive_element.go [hunk 9]
```diff
@@ -32,3 +36,27 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
...
+// Merge "primitive element" is the base case of recurisve call of element merge
+// PrimitiveElement has conflict in following conditions :
+// 1. Recorded and remote values do not match when both of them exist
+// 2. Only one of recorded and remote value exists
+// Example: primitve element list has conflict if value exists in one of recorded/remote but missing in the rest one.
```
Comment (dougm) [inline pkg/kubectl/apply/primitive_element.go]: primitve -> primitive

File: pkg/kubectl/apply/error.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+func (c *ConflictError) Error() string {
+	msg := fmt.Sprintf("[Error] %v : %v", c.Type, c.Message)
+	if c.Detail != "" {
+		msg = fmt.Sprintf("%s, detail: %s", msg, c.Detail)
+	}
+	return msg
+}
+
+// WithDetail adds a detailded message to error
+func (c *ConflictError) WithDetail(format string, v ...interface{}) *ConflictError {
```
[the remark this answers is not in the document]
Comment (dougm) [reply pkg/kubectl/apply/error.go]: Actually, doesn't look like this method is used?

File: pkg/kubectl/apply/primitive_element.go [hunk 11]
```diff
@@ -32,3 +36,27 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
...
+func (e PrimitiveElement) ConflictDetect(v Strategy) (Result, error) {
+	if !checkPrimitive(e) {
+		return Result{}, NewConflictError().WithMessage("Conflict detected in key [%v], recorded value (%+v) and remote value (%+v)",
```
Comment (dougm) [inline pkg/kubectl/apply/primitive_element.go]: typo 'Conflict'

If the `Type` field can be remove are methods `NewConflictError` and `WithMessage` needed, esp. ?  Rather than just:
```go
return Result{}, &ConflictError{Message: fmt.Sprintf("...", ...)}
```

File: pkg/kubectl/apply/element.go [hunk 12]
```diff
@@ -419,3 +419,15 @@ func (e HasElementData) HasLocal() bool {
 func (e HasElementData) HasRemote() bool {
 	return e.remoteSet
 }
+
+// ConflictDetector defines conflict detection method on Element level. Calling ConflictDetect is made by stargey
```
Comment (dougm) [inline pkg/kubectl/apply/element.go]: stargey -> strategy

File: pkg/kubectl/apply/error.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+const (
+	// CONFLICT is type identity for conflicts occurs in recorded and remote
+	CONFLICT = "Conflict"
+)
+
+// ConflictError for conflicts detection, Message is for error message.
+// TODO : Detail field is for more details in error, like full path where conflicts happened.
+type ConflictError struct {
+	Type    string
+	Message string `json:"message"`
```
Comment (dougm) [inline pkg/kubectl/apply/error.go]: Are json tags useful for this type?  If so, `Type` should have a tag if that field is needed (unless Type can be removed of course).

File: pkg/kubectl/apply/parse/list_element.go [hunk 14]
```diff
@@ -127,12 +127,12 @@ func (v ElementBuildingVisitor) doMapList(meta apply.FieldMetaImpl, item *listIt
...
 	}
 
-	for i, l := range orderedKeys.Items {
+	for _, l := range orderedKeys.Items {
 		var s proto.Schema
 		if item.Array != nil && item.Array.SubType != nil {
 			s = item.Array.SubType
 		}
-		subitem, err := v.getItem(s, fmt.Sprintf("%d", i), l.RawElementData)
+		subitem, err := v.getItem(s, fmt.Sprintf("%+v", l.KeyValue), l.RawElementData)
```
Comment (dougm) [inline pkg/kubectl/apply/parse/list_element.go]: Why is this change needed?  In general, not sure about using `%+v` for anything other than logging.  Could it at least be limited to the `PrimitiveListItem.Value` field, to exclude the embedded `RawElementData`?

Comment (fanzhangio) [reply pkg/kubectl/apply/parse/list_element.go]: Aborted this change.

File: pkg/kubectl/apply/strategy/utils_test.go [hunk 15]
```diff
@@ -64,3 +64,22 @@ func create(config string) map[string]interface{} {
...
+		Expect(err).Should(HaveOccurred())
+	} else {
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(merged.Operation).Should(Equal(apply.SET))
+	}
+}
+
+const (
+	hasConflict = true
+	noConflict  = false
```
[the remark this answers is not in the document]
Comment (fanzhangio) [reply pkg/kubectl/apply/strategy/utils_test.go]: They are used in apply/strategy/merge_conflict_test.go, for setting as expected values in tests.

File: pkg/kubectl/apply/error.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+const (
+	// CONFLICT is type identity for conflicts occurs in recorded and remote
+	CONFLICT = "Conflict"
+)
+
+// ConflictError for conflicts detection, Message is for error message.
+// TODO : Detail field is for more details in error, like full path where conflicts happened.
+type ConflictError struct {
+	Type    string
```
[the remark this answers is not in the document]
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: Yes, done.

File: pkg/kubectl/apply/primitive_element.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (fanzhangio) [reply pkg/kubectl/apply/primitive_element.go]: Yes, reflect.DeepEqual is much more reliable.

File: pkg/kubectl/apply/error.go (hunk 13 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: Removed Type field.

File: pkg/kubectl/apply/error.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+// ConflictError for conflicts detection, Message is for error message.
+type ConflictError struct {
+	Message string
+}
+
+// Error implements error
+func (c *ConflictError) Error() string {
+	return fmt.Sprintf("[Error] %v", c.Message)
```
Comment (dougm) [inline pkg/kubectl/apply/error.go]: Just `return c.Message`, no need to add a prefix here.

File: pkg/kubectl/apply/error.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+package apply
+
+import (
+	"fmt"
+)
+
+// ConflictError for conflicts detection, Message is for error message.
+type ConflictError struct {
+	Message string
```
Comment (dougm) [inline pkg/kubectl/apply/error.go]: Good to see this type simplified.  No need to export the `Message` field, since it can be accessed using the `Error` method.

File: pkg/kubectl/apply/list_element.go [hunk 19]
```diff
@@ -65,3 +67,17 @@ func sliceCast(i interface{}) []interface{} {
 	}
 	return i.([]interface{})
 }
+
+// DetectConflict returns ConflictError if fields in recorded and remote of ListElement conflict
+// ConflictError Message contains details of conflict.
+func (e ListElement) DetectConflict() error {
+	for _, item := range e.Values {
+		if IsConflict(item) {
+			return &ConflictError{Message: fmt.Sprintf("Conflict detected in list elements, recorded value (%+v) and remote value (%+v)",
```
Comment (dougm) [inline pkg/kubectl/apply/list_element.go]: nit, but the message says `list elements` here, rather than `ListElement`.  A later message below uses `MapElement`.
Keeping these messages consistent and code a bit more concise would be a good case for something like this helper:

```go
func newConflictError(e Element) *ConflictError {
	return &ConflictError{
		message: fmt.Sprintf("conflict detected in %s, recorded value (%+v) and remote value (%+v)",
			e.GetFieldType(), e.GetRecorded(), e.GetRemote()),
	}
}
```

Also note `conflict` rather than `Conflict`: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: pkg/kubectl/apply/element.go [hunk 20]
```diff
@@ -325,6 +326,18 @@ func IsAdd(e Element) bool {
 	return e.HasLocal() && !e.HasRemote()
 }
 
+// IsConflict returns true if the object represented by e should either have the same field specified
```
Comment (droot) [inline pkg/kubectl/apply/element.go]: We can simplify the comment by saying:
"returns true if the field represented by e has conflict between remote and recorded value".

File: pkg/kubectl/apply/element.go [hunk 21]
```diff
@@ -419,3 +432,11 @@ func (e HasElementData) HasLocal() bool {
 func (e HasElementData) HasRemote() bool {
 	return e.remoteSet
 }
+
+// DetectConflictor defines conflict detection method on Element level.
+// Calling DetectConflict on an element will apply type specific detection method
+// A conflict is if the same field is specified in BOTH the recorded and the remote values
+// of an object, but does not match.
+type DetectConflictor interface {
```
Comment (droot) [inline pkg/kubectl/apply/element.go]: We should rename it to ConflictDetector. Its defining behavior for "detecting conflicts", doesn't have follow "xxxx-er" like syntax :) I know where you are coming from :)

We should improve the documentation saying:
// ConflictDetector defines the capability to detect conflict. An element can examine remote/recorded value to detect conflict. 
Also: we don't have to repeat definition of conflict everywhere. We can just define it at one place.

File: pkg/kubectl/apply/list_element.go [hunk 22]
```diff
@@ -65,3 +67,19 @@ func sliceCast(i interface{}) []interface{} {
 	}
 	return i.([]interface{})
 }
+
+// DetectConflict returns ConflictError if fields in recorded and remote of ListElement conflict
+// ConflictError Message contains details of conflict.
```
Comment (droot) [inline pkg/kubectl/apply/list_element.go]: we can remove the second comment because that is captured in definition of the ConflictError.

File: pkg/kubectl/apply/map_element.go [hunk 23]
```diff
@@ -70,3 +72,19 @@ func mapCast(i interface{}) map[string]interface{} {
 	}
 	return i.(map[string]interface{})
 }
+
+// DetectConflict returns ConflictError if some elements in map conflict.
+// ConflictError Message contains details of conflict.
```
Comment (droot) [inline pkg/kubectl/apply/map_element.go]: kill the second comment.

File: pkg/kubectl/apply/strategy/strategic_visitor.go [hunk 24]
```diff
@@ -96,4 +96,16 @@ func (v delegatingStrategy) MergeEmpty(diff apply.EmptyElement) (apply.Result, e
 	return v.merge.MergeEmpty(diff)
 }
 
+// DoConflictDetect detects conflicts in element when option enabled, return error if conflict happened.
+func (v delegatingStrategy) DoConflictDetect(e apply.Element) error {
+	if v.options.FailOnConflict {
+		if e, ok := e.(apply.DetectConflictor); ok {
+			if err := e.DetectConflict(); err != nil {
```
Comment (droot) [inline pkg/kubectl/apply/strategy/strategic_visitor.go]: return e.DetectConflict() directly ?

File: pkg/kubectl/apply/visitor.go [hunk 25]
```diff
@@ -66,3 +66,8 @@ const (
 	// ReplaceStrategy is the strategy to replace the remote value with the local value
 	ReplaceStrategy = "replace"
 )
+
+// DoConflictDetector on strategy for element calling conflict method
+type DoConflictDetector interface {
```
Comment (droot) [inline pkg/kubectl/apply/visitor.go]: wondering why do we have this interface ?

**Review by droot — COMMENTED** [review]
Fan, Thanks for iterating on it. I think its looking better.

I made a first pass and have a few comments. Will be taking another look to complete the review. 

Also would recommend running godoc on your code and review the documentation in the browser to see how the comments will read out to an external reader. I have found that to be super useful for documentation review.

File: pkg/kubectl/apply/visitor.go (hunk 25 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/visitor.go]: This interface is used for abstraction of defining a generic behavior on strategy level. It invokes type specific conflict method of element. Whatever merge strategy, replace strategy or delegate strategy has this kind of similar behavior. I want to keep the code style following the "visitor pattern" in merge and replace code base without breaking out. If conflict detection logic needs to be iterated in the future, the DoConflictDetector implement is the only place needs to be changed.

File: pkg/kubectl/apply/error.go [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+// ConflictError for conflicts detection, Message is for error message.
+type ConflictError struct {
+	message string
```
[the remark this answers is not in the document]
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: Good point!

File: pkg/kubectl/apply/element.go (hunk 20 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/element.go]: nice!

File: pkg/kubectl/apply/strategy/strategic_visitor.go (hunk 24 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/strategy/strategic_visitor.go]: Yes, that's a typo :P

File: pkg/kubectl/apply/visitor.go (hunk 25 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/visitor.go]: It is used for an abstraction of a generic behavior on strategy level. This behavior invokes element's type specific conflict detection method for each coming element. All strategies, like merge and replace, will do the behavior in the similar way, so calling the interface to encapsulate the underlying implementation for each. Another benefit is that it retains the merge and replace strategy code base following "visitor" pattern. If there are iterations in conflict detection behaviors, the DoConflictDetector implementation is the only place needs to be changed.

File: pkg/kubectl/apply/error.go [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+// ConflictError for conflicts detection, Message is for error message.
```
[the remark this answers is not in the document]
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: It looks like self-defined error struct is named like "XXXError", referring to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/error.go
and https://blog.golang.org/error-handling-and-go

File: pkg/kubectl/apply/visitor.go (hunk 25 above)
Comment (anfernee) [reply pkg/kubectl/apply/visitor.go]: I guess the question is why you have this interface, while it's not used anywhere.

File: pkg/kubectl/apply/error.go [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+package apply
+
+import "fmt"
+
+// ConflictError represents a conflict error occurred during the merge operation.
+type ConflictError struct {
+	message string
+}
+
+// NewConflictError returns a ConflictError with detailed conflict informaion in element
```
Comment (anfernee) [inline pkg/kubectl/apply/error.go]: informaion => information

File: pkg/kubectl/apply/strategy/merge_visitor.go [hunk 29]
```diff
@@ -41,7 +41,10 @@ func (v mergeStrategy) MergeList(e apply.ListElement) (apply.Result, error) {
 	if result, done := v.doAddOrDelete(e); done {
 		return result, nil
 	}
-
```
Comment (anfernee) [inline pkg/kubectl/apply/strategy/merge_visitor.go]: I am not sure about this, but why not doing this before `doAddOrDelete`. Same for others. also it might be a good idea to change it to `doConflictDetect`, I assume it's internal.

File: pkg/kubectl/apply/strategy/merge_visitor.go [hunk 30]
```diff
@@ -145,4 +154,9 @@ func (v mergeStrategy) MergeEmpty(diff apply.EmptyElement) (apply.Result, error)
 	return apply.Result{Operation: apply.SET}, nil
 }
 
+// DoConflictDetect implement on mergeStrategy to return error if element has conflict
```
Comment (anfernee) [inline pkg/kubectl/apply/strategy/merge_visitor.go]: as mentioned, make it `doConflictDetect`

File: pkg/kubectl/apply/strategy/utils_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (anfernee) [reply pkg/kubectl/apply/strategy/utils_test.go]: can you move them to where they are used.

**Review by anfernee — COMMENTED** [review]
This looks very close to be approved. some nits.

File: pkg/kubectl/apply/visitor.go (hunk 25 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/visitor.go]: Yes, interface abstraction is not necessary. Just define and implement method on strategy is sufficient

File: pkg/kubectl/apply/error.go (hunk 28 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: done

File: pkg/kubectl/apply/strategy/merge_visitor.go (hunk 29 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/strategy/merge_visitor.go]: I believe that doAddOrDelete() has no merge logic if adding or deleting. It should go prior to merge. If "not done" returns from doAddOrDelete(), it starts to merge and conflict detection if optioned.

File: pkg/kubectl/apply/element.go [hunk 31]
```diff
@@ -325,6 +326,17 @@ func IsAdd(e Element) bool {
...
 
+// IsConflict returns true if the field represented by e has conflict between remote and recorded value
+func IsConflict(e Element) bool {
+	if !e.HasRecorded() && !e.HasRemote() {
+		return false
+	}
+	if e.HasRecorded() && e.HasRemote() {
+		return !reflect.DeepEqual(e.GetRecorded(), e.GetRemote())
+	}
+	return true
```
Comment (pwittrock) [inline pkg/kubectl/apply/element.go]: If there is a Remote and !Recorded it is a conflict?

File: pkg/kubectl/apply/element.go [hunk 32]
```diff
@@ -325,6 +326,17 @@ func IsAdd(e Element) bool {
 	return e.HasLocal() && !e.HasRemote()
 }
 
+// IsConflict returns true if the field represented by e has conflict between remote and recorded value
```
Comment (pwittrock) [inline pkg/kubectl/apply/element.go]: Document what qualifies as a conflict and why;

File: pkg/kubectl/apply/element.go [hunk 33]
```diff
@@ -325,6 +326,17 @@ func IsAdd(e Element) bool {
...
+func IsConflict(e Element) bool {
+	if !e.HasRecorded() && !e.HasRemote() {
+		return false
+	}
+	if e.HasRecorded() && e.HasRemote() {
+		return !reflect.DeepEqual(e.GetRecorded(), e.GetRemote())
```
Comment (pwittrock) [inline pkg/kubectl/apply/element.go]: We should only do this check on PrimitiveElements, it doesn't make sense for maps or lists

File: pkg/kubectl/apply/primitive_element.go [hunk 34]
```diff
@@ -32,3 +32,13 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
 }
 
 var _ Element = &PrimitiveElement{}
+
+// DetectConflict returns ConflictError if primitive element has conflict field.
+func (e PrimitiveElement) DetectConflict() error {
+	if IsConflict(e) {
+		return NewConflictError(e)
```
Comment (pwittrock) [inline pkg/kubectl/apply/primitive_element.go]: Inline the IsConflict function here, nothing else should use it

File: pkg/kubectl/apply/map_element.go [hunk 35]
```diff
@@ -70,3 +70,15 @@ func mapCast(i interface{}) map[string]interface{} {
 	}
 	return i.(map[string]interface{})
 }
+
+// DetectConflict returns ConflictError if some elements in map conflict.
+func (e MapElement) DetectConflict() error {
+	for _, item := range e.GetValues() {
+		if IsConflict(item) {
```
Comment (pwittrock) [inline pkg/kubectl/apply/map_element.go]:
```go

if err := item.DetectConflict(); err != nil {
  return err
}

File: pkg/kubectl/apply/list_element.go [hunk 36]
```diff
@@ -65,3 +65,15 @@ func sliceCast(i interface{}) []interface{} {
 	}
 	return i.([]interface{})
 }
+
+// DetectConflict returns ConflictError if fields in recorded and remote of ListElement conflict
+func (e ListElement) DetectConflict() error {
+	for _, item := range e.Values {
+		if IsConflict(item) {
```
Comment (pwittrock) [inline pkg/kubectl/apply/list_element.go]:
```go
if err := item.DetectConflict(); err != nil {
  return err
}
```

File: pkg/kubectl/apply/element.go [hunk 37]
```diff
@@ -419,3 +431,8 @@ func (e HasElementData) HasLocal() bool {
 func (e HasElementData) HasRemote() bool {
 	return e.remoteSet
 }
+
+// ConflictDetector defines the capability to detect conflict. An element can examine remote/recorded value to detect conflict.
+type ConflictDetector interface {
+	DetectConflict() error
```
Comment (pwittrock) [inline pkg/kubectl/apply/element.go]:
```go
HasConflict() error
```

File: pkg/kubectl/apply/error.go [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+
+import "fmt"
+
+// ConflictError represents a conflict error occurred during the merge operation.
+type ConflictError struct {
+	message string
+}
+
+// NewConflictError returns a ConflictError with detailed conflict information in element
+func NewConflictError(e Element) *ConflictError {
```
Comment (pwittrock) [inline pkg/kubectl/apply/error.go]: This should probably be PrimitiveElement only

File: pkg/kubectl/apply/error.go [hunk 39]
```diff
@@ -0,0 +1,37 @@
...
+	return &ConflictError{
+		message: fmt.Sprintf("conflict detected in %s, recorded value (%+v) and remote value (%+v)",
+			e.GetFieldType(), e.GetRecorded(), e.GetRemote()),
+	}
+}
+
+// Error implements error
+func (c *ConflictError) Error() string {
```
Comment (pwittrock) [inline pkg/kubectl/apply/error.go]: Is there a way we can make it so we can add to the field path?

File: pkg/kubectl/apply/error.go [hunk 40]
```diff
@@ -0,0 +1,37 @@
...
+
+// ConflictError represents a conflict error occurred during the merge operation.
+type ConflictError struct {
+	message string
+}
+
+// NewConflictError returns a ConflictError with detailed conflict information in element
+func NewConflictError(e Element) *ConflictError {
+	return &ConflictError{
+		message: fmt.Sprintf("conflict detected in %s, recorded value (%+v) and remote value (%+v)",
```
Comment (pwittrock) [inline pkg/kubectl/apply/error.go]: Might want to put this data in the struct and generate the string message when needed

**Comment by fanzhangio** [comment]
kind/feature

File: pkg/kubectl/apply/error.go (hunk 39 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: Yes, could we take this feature in a separate issue ?https://github.com/kubernetes/kubernetes/issues/61437

File: pkg/kubectl/apply/element.go (hunk 37 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/element.go]: I think normally **Has...** method returns a boolean, while DetectConflict() which is a executive behavior returns error. Dose it make sense?

File: pkg/kubectl/apply/error.go (hunk 38 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/error.go]: Make sense!

**Review by soltysh — APPROVED** [review]

**Comment by soltysh** [comment]
/hold
The code itself lgtm here, but I'd like to hear from @lavalamp and @liggitt based on [this comment](https://github.com/kubernetes/kubernetes/issues/60945#issuecomment-371692213).

**Comment by liggitt** [comment]
My main concern is investing a lot of new development and review time in a feature with an uncertain future. We probably should not expand the API surface of the current apply and strategic merge patch features until the server-side direction is resolved. If this doesn't add new API surface area, I don't object to it. 

Will defer to sig-cli/sig-api-machinery to decide if this advances in a direction that is helpful long term.

File: pkg/kubectl/apply/primitive_element.go [hunk 41]
```diff
@@ -32,3 +34,24 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
...
+// 1. A field is specified in both recorded and remote values, but does not match.
+// 2. A field is specified in recorded values, but missing in remote values.
+func (e PrimitiveElement) HasConflict() error {
+	if !e.HasRecorded() && !e.HasRemote() {
+		return nil
+	}
+	if e.HasRecorded() && e.HasRemote() {
+		if !reflect.DeepEqual(e.GetRecorded(), e.GetRemote()) {
+			return NewConflictError(e)
+		}
```
Comment (droot) [inline pkg/kubectl/apply/primitive_element.go]: shell we return nil from here itself ? because has recorded and remote and are equal ?

File: pkg/kubectl/apply/primitive_element.go [hunk 42]
```diff
@@ -32,3 +34,24 @@ func (e PrimitiveElement) Merge(v Strategy) (Result, error) {
...
 var _ Element = &PrimitiveElement{}
+
+// HasConflict returns ConflictError if primitive element has conflict field.
+// Conflicts happen when either of the following conditions:
+// 1. A field is specified in both recorded and remote values, but does not match.
+// 2. A field is specified in recorded values, but missing in remote values.
+func (e PrimitiveElement) HasConflict() error {
+	if !e.HasRecorded() && !e.HasRemote() {
+		return nil
+	}
```
Comment (droot) [inline pkg/kubectl/apply/primitive_element.go]: we can take this `if` block out now. It will fall back to the last `return nil` statement and will be consistent with the documentation as well :)

**Review by droot — COMMENTED** [review]
Looks good to me. Minor suggestions.

**Comment by droot** [comment]
@liggitt @soltysh These changes do not change the API surface. They are just improving the error handling part by adding conflict detection. Also note that this `apply` package is only being used by `kubectl alpha diff` not the existing apply command yet. This pkg can potentially be used in the server side Apply effort.

File: pkg/kubectl/apply/primitive_element.go (hunk 42 above)
Comment (fanzhangio) [reply pkg/kubectl/apply/primitive_element.go]: Yes, it should be cleaned so that the code logic is aline with documentation.

**Comment by lavalamp** [comment]
I don't object to changing the existing apply, provided that:
* No API changes
* it is evolving to behave more like the planned apply rather than less like it

**Review by soltysh — APPROVED** [review]
/lgtm
/approve no-issue
/hold cancel
